File: crates/loop/ironclaw_loop_host/src/remote_host/protocol.rs
use ironclaw_host_api::process::MAX_SANDBOX_LOOP_WORKER_FRAME_BYTES;
use ironclaw_loop_contracts::*;
use ironclaw_turns::LoopMessageRef;
use serde::{Deserialize, Serialize, de::DeserializeOwned};

pub const LOOP_WORKER_WIRE_VERSION: u16 = 2;
pub const LOOP_WORKER_MAX_FRAME_BYTES: usize = MAX_SANDBOX_LOOP_WORKER_FRAME_BYTES;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum LoopWorkerInvocation {
    Run(AgentLoopDriverRunRequest),
    Resume(AgentLoopDriverResumeRequest),
}

#[derive(Debug, Clone, Copy, Default, Serialize, Deserialize)]
pub struct LoopWorkerSettings {
    pub default_iteration_limit: Option<u32>,
    pub model_availability_attempts: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LoopWorkerBootstrap {
    pub wire_version: u16,
    pub run_context: LoopRunContext,
    pub invocation: LoopWorkerInvocation,
    pub settings: LoopWorkerSettings,
    pub tool_definitions: Vec<ProviderToolDefinition>,
    pub current_visible_capabilities: Option<serde_json::Value>,
    /// Whether the worker is allowed to see resolved transcript content.
    /// An omitted field is blind; workers still reject unsupported wire versions.
    #[serde(default)]
    pub content_visibility: WorkerContentVisibility,
}

/// Whether the worker was bootstrapped allowed to request resolved
/// model-visible content for the message refs the host already issued.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WorkerContentVisibility {
    /// The worker sees refs and safe summaries only (canonical Rust worker).
    #[default]
    Blind,
    /// The worker may call `HostCall::ResolveMessages` for issued refs.
    Resolved,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LoopWorkerFailure {
    pub kind: String,
    pub detail: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum LoopWorkerOutcome {
    Exit(LoopExit),
    Failed(LoopWorkerFailure),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HostRequestFrame {
    pub id: u64,
    pub call: HostCall,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HostResponseFrame {
    pub id: u64,
    pub result: Result<serde_json::Value, WireError>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum WorkerFrame {
    HostRequest(Box<HostRequestFrame>),
    Outcome(LoopWorkerOutcome),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum HostFrame {
    Bootstrap(Box<LoopWorkerBootstrap>),
    HostResponse(HostResponseFrame),
    Cancel(LoopCancellationSignal),
    OutcomeAck,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum WireError {
    Host(AgentLoopHostError),
    Compaction(LoopCompactionError),
    Protocol(String),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum HostCall {
    ResolveMessages(ResolveMessagesRequest),
    LoadContext(LoopContextRequest),
    BuildPrompt(LoopPromptBundleRequest),
    PollInputs {
        after: LoopInputCursor,
        limit: usize,
    },
    AckInputs(Vec<LoopInputAckToken>),
    StreamModel(LoopModelRequest),
    RegisterProviderToolCall(RegisterProviderToolCallRequest),
    VisibleCapabilities(VisibleCapabilityRequest),
    InvokeCapability(LoopRequest),
    InvokeCapabilityBatch(LoopRequestBatch),
    BeginAssistantDraft(BeginAssistantDraft),
    UpdateAssistantDraft(UpdateAssistantDraft),
    FinalizeAssistantMessage(FinalizeAssistantMessage),
    AppendCapabilityResultRef(Box<AppendCapabilityResultRef>),
    Checkpoint(LoopCheckpointRequest),
    StageCheckpointPayload(StageCheckpointPayloadRequest),
    LoadCheckpointPayload(LoadCheckpointPayloadRequest),
    EmitProgress(LoopProgressEvent),
    Compact(LoopCompactionRequest),
}

/// Request for host-resolved model-visible content of message refs the run
/// already holds. Only honored for a worker bootstrapped `Resolved`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ResolveMessagesRequest {
    pub messages: Vec<LoopModelMessage>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct WireResolvedToolResult {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub provider_call_id: Option<String>,
    pub content: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct WireResolvedModelMessage {
    pub role: String,
    pub content_ref: LoopMessageRef,
    pub content: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tool_result: Option<WireResolvedToolResult>,
}

pub(super) fn encode<T: Serialize>(value: &T) -> Result<Vec<u8>, AgentLoopHostError> {
    let bytes = serde_json::to_vec(value).map_err(|error| {
        AgentLoopHostError::new(
            AgentLoopHostErrorKind::Internal,
            format!("loop worker frame serialization failed: {error}"),
        )
    })?;
    if bytes.len() > LOOP_WORKER_MAX_FRAME_BYTES {
        return Err(AgentLoopHostError::new(
            AgentLoopHostErrorKind::InvalidInvocation,
            "loop worker frame exceeds the configured byte limit",
        ));
    }
    Ok(bytes)
}

pub(super) fn decode<T: DeserializeOwned>(bytes: &[u8]) -> Result<T, AgentLoopHostError> {
    if bytes.len() > LOOP_WORKER_MAX_FRAME_BYTES {
        return Err(AgentLoopHostError::new(
            AgentLoopHostErrorKind::InvalidInvocation,
            "loop worker frame exceeds the configured byte limit",
        ));
    }
    serde_json::from_slice(bytes).map_err(|error| {
        AgentLoopHostError::new(
            AgentLoopHostErrorKind::InvalidInvocation,
            format!("loop worker frame is malformed: {error}"),
        )
    })
}
#[derive(Debug, Clone, Serialize, Deserialize)]
pub(super) struct WireLoopContextBundle {
    identity_messages: Vec<WireLoopContextMessage>,
    messages: Vec<WireLoopContextMessage>,
    compaction_message_index: Vec<LoopContextCompactionMetadata>,
    recent_window_truncation: Option<LoopContextWindowTruncation>,
    instruction_snippets: Vec<WireLoopContextSnippet>,
    memory_snippets: Vec<WireLoopContextSnippet>,
}
#[derive(Debug, Clone, Serialize, Deserialize)]
pub(super) struct WireLoadedCheckpointPayload {
    kind: LoopCheckpointKind,
    schema_id: CheckpointSchemaId,
    schema_version: ironclaw_host_api::turn::RunProfileVersion,
    payload: Vec<u8>,
}

impl From<LoadedCheckpointPayload> for WireLoadedCheckpointPayload {
    fn from(payload: LoadedCheckpointPayload) -> Self {
        Self {
            kind: payload.kind,
            schema_id: payload.schema_id,
            schema_version: payload.schema_version,
            payload: payload.payload.into_payload_bytes(),
        }
    }
}

impl TryFrom<WireLoadedCheckpointPayload> for LoadedCheckpointPayload {
    type Error = AgentLoopHostError;

    fn try_from(payload: WireLoadedCheckpointPayload) -> Result<Self, Self::Error> {
        Ok(Self {
            kind: payload.kind,
            schema_id: payload.schema_id,
            schema_version: payload.schema_version,
            payload: RedactedCheckpointPayload::new(payload.payload).map_err(|error| {
                AgentLoopHostError::new(
                    AgentLoopHostErrorKind::InvalidInvocation,
                    format!("remote checkpoint payload is invalid: {error}"),
                )
            })?,
        })
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub(super) struct WireVisibleCapabilitySurface {
    version: CapabilitySurfaceVersion,
    descriptors: Vec<WireCapabilityDescriptorView>,
    callable_capability_ids: Option<Vec<ironclaw_host_api::ids::CapabilityId>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub(super) struct WireCapabilityDescriptorView {
    capability_id: ironclaw_host_api::ids::CapabilityId,
    provider: Option<ironclaw_host_api::ids::ExtensionId>,
    #[serde(deserialize_with = "ironclaw_host_api::runtime::deserialize_trusted_runtime_kind")]
    runtime: ironclaw_host_api::runtime::RuntimeKind,
    safe_name: String,
    safe_description: String,
    description_trust: ironclaw_host_api::capability::CapabilityDescriptionTrust,
    parameters_schema: serde_json::Value,
}

impl From<VisibleCapabilitySurface> for WireVisibleCapabilitySurface {
    fn from(surface: VisibleCapabilitySurface) -> Self {
        Self {
            version: surface.version,
            descriptors: surface
                .descriptors
                .into_iter()
                .map(WireCapabilityDescriptorView::from)
                .collect(),
            callable_capability_ids: surface.callable_capability_ids,
        }
    }
}

impl From<WireVisibleCapabilitySurface> for VisibleCapabilitySurface {
    fn from(surface: WireVisibleCapabilitySurface) -> Self {
        Self {
            version: surface.version,
            descriptors: surface
                .descriptors
                .into_iter()
                .map(CapabilityDescriptorView::from)
                .collect(),
            callable_capability_ids: surface.callable_capability_ids,
        }
    }
}

impl From<CapabilityDescriptorView> for WireCapabilityDescriptorView {
    fn from(descriptor: CapabilityDescriptorView) -> Self {
        Self {
            capability_id: descriptor.capability_id,
            provider: descriptor.provider,
            runtime: descriptor.runtime,
            safe_name: descriptor.safe_name,
            safe_description: descriptor.safe_description,
            description_trust: descriptor.description_trust,
            parameters_schema: descriptor.parameters_schema,
        }
    }
}

impl From<WireCapabilityDescriptorView> for CapabilityDescriptorView {
    fn from(descriptor: WireCapabilityDescriptorView) -> Self {
        Self {
            capability_id: descriptor.capability_id,
            provider: descriptor.provider,
            runtime: descriptor.runtime,
            safe_name: descriptor.safe_name,
            safe_description: descriptor.safe_description,
            description_trust: descriptor.description_trust,
            parameters_schema: descriptor.parameters_schema,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub(super) struct WireLoopContextMessage {
    message_ref: Option<ironclaw_host_api::turn::LoopMessageRef>,
    role: String,
    safe_summary: String,
    compaction: Option<LoopContextCompactionMetadata>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub(super) struct WireLoopContextSnippet {
    snippet_ref: String,
    model_content: String,
    safe_summary: String,
    metadata: Option<WireLoopContextSnippetMetadata>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub(super) struct WireLoopContextSnippetMetadata {
    source_name: String,
    trust_level: SkillTrustLevel,
}

impl From<LoopContextBundle> for WireLoopContextBundle {
    fn from(bundle: LoopContextBundle) -> Self {
        Self {
            identity_messages: bundle
                .identity_messages
                .into_iter()
                .map(WireLoopContextMessage::from)
                .collect(),
            messages: bundle
                .messages
                .into_iter()
                .map(WireLoopContextMessage::from)
                .collect(),
            compaction_message_index: bundle.compaction_message_index,
            recent_window_truncation: bundle.recent_window_truncation,
            instruction_snippets: bundle
                .instruction_snippets
                .into_iter()
                .map(WireLoopContextSnippet::from)
                .collect(),
            memory_snippets: bundle
                .memory_snippets
                .into_iter()
                .map(WireLoopContextSnippet::from)
                .collect(),
        }
    }
}

impl From<WireLoopContextBundle> for LoopContextBundle {
    fn from(bundle: WireLoopContextBundle) -> Self {
        Self {
            identity_messages: bundle
                .identity_messages
                .into_iter()
                .map(LoopContextMessage::from)
                .collect(),
            messages: bundle
                .messages
                .into_iter()
                .map(LoopContextMessage::from)
                .collect(),
            compaction_message_index: bundle.compaction_message_index,
            recent_window_truncation: bundle.recent_window_truncation,
            instruction_snippets: bundle
                .instruction_snippets
                .into_iter()
                .map(LoopContextSnippet::from)
                .collect(),
            memory_snippets: bundle
                .memory_snippets
                .into_iter()
                .map(LoopContextSnippet::from)
                .collect(),
        }
    }
}

impl From<LoopContextMessage> for WireLoopContextMessage {
    fn from(message: LoopContextMessage) -> Self {
        Self {
            message_ref: message.message_ref,
            role: message.role,
            safe_summary: message.safe_summary,
            compaction: message.compaction,
        }
    }
}

impl From<WireLoopContextMessage> for LoopContextMessage {
    fn from(message: WireLoopContextMessage) -> Self {
        Self {
            message_ref: message.message_ref,
            role: message.role,
            safe_summary: message.safe_summary,
            compaction: message.compaction,
        }
    }
}

impl From<LoopContextSnippet> for WireLoopContextSnippet {
    fn from(snippet: LoopContextSnippet) -> Self {
        Self {
            snippet_ref: snippet.snippet_ref,
            model_content: snippet.model_content,
            safe_summary: snippet.safe_summary,
            metadata: snippet.metadata.map(WireLoopContextSnippetMetadata::from),
        }
    }
}

impl From<WireLoopContextSnippet> for LoopContextSnippet {
    fn from(snippet: WireLoopContextSnippet) -> Self {
        Self {
            snippet_ref: snippet.snippet_ref,
            model_content: snippet.model_content,
            safe_summary: snippet.safe_summary,
            metadata: snippet.metadata.map(LoopContextSnippetMetadata::from),
        }
    }
}

impl From<LoopContextSnippetMetadata> for WireLoopContextSnippetMetadata {
    fn from(metadata: LoopContextSnippetMetadata) -> Self {
        Self {
            source_name: metadata.source_name,
            trust_level: metadata.trust_level,
        }
    }
}

impl From<WireLoopContextSnippetMetadata> for LoopContextSnippetMetadata {
    fn from(metadata: WireLoopContextSnippetMetadata) -> Self {
        Self {
            source_name: metadata.source_name,
            trust_level: metadata.trust_level,
        }
    }
}

File: docker/sandbox/pi-worker/src/protocol.ts
/**
 * TypeScript mirror of the IronClaw loop-worker wire protocol
 * (`crates/loop/ironclaw_loop_host/src/remote_host/protocol.rs`, wire v2).
 *
 * Every type here matches serde's JSON output exactly: snake_case fields,
 * externally tagged enums (`{"HostRequest": {...}}`), and Rust `Result` as
 * `{"Ok": ...}` / `{"Err": ...}`. Opaque host-owned values (checkpoint bytes,
 * capability views, run profiles) are typed `unknown` and passed through
 * verbatim — the worker never interprets them.
 */

/** `LOOP_WORKER_WIRE_VERSION` — v2 adds `content_visibility` and `ResolveMessages`. */
export const LOOP_WORKER_WIRE_VERSION = 2;

/** `LOOP_WORKER_MAX_FRAME_BYTES` = `MAX_SANDBOX_LOOP_WORKER_FRAME_BYTES`. */
export const LOOP_WORKER_MAX_FRAME_BYTES = 1024 * 1024;

// ---------------------------------------------------------------------------
// Bootstrap
// ---------------------------------------------------------------------------

/** `WorkerContentVisibility` — serde `rename_all = "snake_case"`. */
export type WorkerContentVisibility = "blind" | "resolved";

/**
 * `LoopWorkerBootstrap`. Opaque fields (`run_context`, `invocation`,
 * `current_visible_capabilities`) are passed through untouched; only the
 * fields the worker must act on are narrowed.
 */
export interface LoopWorkerBootstrap {
  wire_version: number;
  run_context: LoopRunContext;
  invocation: LoopWorkerInvocation;
  settings: LoopWorkerSettings;
  tool_definitions: ProviderToolDefinition[];
  current_visible_capabilities: unknown | null;
  /** v2; serde default `blind` so v1-shaped bootstraps decode. */
  content_visibility?: WorkerContentVisibility;
}

export interface LoopWorkerSettings {
  default_iteration_limit: number | null;
  model_availability_attempts: number | null;
}

/**
 * `LoopWorkerInvocation { Run(...)|Resume(...) }` — externally tagged.
 * The request payloads (`AgentLoopDriverRunRequest` / `ResumeRequest`) are
 * opaque to the worker: the host drives scoping and the worker never reads
 * `resolved_run_profile` (the host enforces budgets itself).
 */
export type LoopWorkerInvocation =
  { Run: AgentLoopDriverRunRequest } | { Resume: AgentLoopDriverResumeRequest };

/** `AgentLoopDriverRunRequest` shape, kept for the conformance fake host. */
export interface AgentLoopDriverRunRequest {
  turn_id: string;
  run_id: string;
  resolved_run_profile: unknown;
}

/** `AgentLoopDriverResumeRequest` shape, kept for the conformance fake host. */
export interface AgentLoopDriverResumeRequest {
  turn_id: string;
  run_id: string;
  checkpoint_id: string;
  resolved_run_profile: unknown;
  auth_resume_disposition?: "denied" | null;
}

// ---------------------------------------------------------------------------
// Loop context (mirrors `host/run_context.rs`; opaque pass-through)
// ---------------------------------------------------------------------------

/** `LoopRunContext` — opaque to the worker; echoed in `BuildPrompt`. */
export interface LoopRunContext {
  scope: unknown;
  actor?: unknown;
  accepted_message_ref?: unknown;
  thread_id: string;
  turn_id: string;
  run_id: string;
  resolved_run_profile: unknown;
  resolved_model_route?: unknown;
  loop_driver_id: string;
  loop_driver_version: number;
  checkpoint_schema_id: string;
  checkpoint_schema_version: number;
  output_contract?: unknown;
  product_context?: unknown;
}

// ---------------------------------------------------------------------------
// Host calls the worker may issue (subset it uses)
// ---------------------------------------------------------------------------

/** `ProviderToolDefinition` (`description_trust` is `#[serde(skip)]`). */
export interface ProviderToolDefinition {
  capability_id: string;
  name: string;
  description: string;
  parameters: unknown;
}

/**
 * `HostCall` — externally tagged. Only the calls the Pi worker issues are
 * typed precisely; the rest stay `unknown` payloads.
 */
export type HostCall =
  | { BuildPrompt: LoopPromptBundleRequest }
  | { ResolveMessages: ResolveMessagesRequest }
  | { StreamModel: LoopModelRequest }
  | { RegisterProviderToolCall: RegisterProviderToolCallRequest }
  | { InvokeCapability: LoopRequest }
  | { BeginAssistantDraft: BeginAssistantDraft }
  | { UpdateAssistantDraft: UpdateAssistantDraft }
  | { FinalizeAssistantMessage: FinalizeAssistantMessage }
  | { AppendCapabilityResultRef: AppendCapabilityResultRef }
  | { StageCheckpointPayload: StageCheckpointPayloadRequest }
  | { EmitProgress: unknown }
  | { [variant: string]: unknown };

// --- prompt bundle (host/model.rs) ---

/** `PromptMode` — serde snake_case. */
export type PromptMode = "text_only" | "codeact";

/**
 * `LoopInlineMessageRole` — serde snake_case. Leading messages must stay in
 * the provider-cached leading block.
 */
export type LoopInlineMessageRole =
  "leading_system" | "leading_user" | "system" | "user" | "assistant";

export interface LoopInlineMessage {
  role: LoopInlineMessageRole;
  safe_body: string;
}

export interface LoopPromptBundleRequest {
  mode: PromptMode;
  context_cursor: LoopInputCursor | null;
  surface_version: string | null;
  capability_view?: LoopModelCapabilityView | null;
  checkpoint_state_ref: string | null;
  max_messages: number | null;
  inline_messages?: LoopInlineMessage[];
}

export interface LoopInputCursor {
  scope: unknown;
  run_id: string;
  token: string;
}

export interface LoopModelCapabilityView {
  visible_capability_ids: string[];
}

/** `LoopPromptBundle` — the host's prompt bundle response. */
export interface LoopPromptBundle {
  bundle_ref: string;
  messages: LoopModelMessage[];
  surface_version: string | null;
  compaction_message_index?: LoopContextCompactionMetadata[];
  recent_window_truncation?: unknown;
  instruction_fingerprint?: unknown;
  identity_message_count: number;
  instruction_snippet_count: number;
}

export interface LoopContextCompactionMetadata {
  [key: string]: unknown;
}

// --- message refs and resolution (v2) ---

/** `LoopModelMessage` — role + opaque ref. */
export interface LoopModelMessage {
  role: string;
  content_ref: string;
}

/** v2 `ResolveMessages` request. */
export interface ResolveMessagesRequest {
  messages: LoopModelMessage[];
}

/** v2 `WireResolvedToolResult`. */
export interface WireResolvedToolResult {
  provider_call_id: string | null;
  content: string;
}

/** v2 `WireResolvedModelMessage`. */
export interface WireResolvedModelMessage {
  role: string;
  content_ref: string;
  content: string;
  tool_result?: WireResolvedToolResult | null;
}

// --- model call ---

export interface LoopModelRequest {
  messages: LoopModelMessage[];
  inline_messages?: LoopInlineMessage[];
  surface_version: string | null;
  model_preference: string | null;
  fallback_index: number;
  iteration: number;
  capability_view?: LoopModelCapabilityView | null;
  tool_choice?: LoopModelToolChoice | null;
}

/** `LoopModelToolChoice` — serde `tag = "kind"`, snake_case. */
export type LoopModelToolChoice = {
  kind: "forced_capability";
  capability_id: string;
};

/** `LoopModelResponse` returned by `StreamModel`. */
export interface LoopModelResponse {
  chunks: ModelStreamChunk[];
  safe_reasoning_deltas?: string[];
  output: ParentLoopOutput;
  effective_model_profile_id: string;
  usage?: LoopModelUsage | null;
}

export interface ModelStreamChunk {
  safe_text_delta: string;
}

/** `ParentLoopOutput` — externally tagged. */
export type ParentLoopOutput =
  | { assistant_reply: AssistantReply }
  | { capability_calls: CapabilityCallCandidate[] };

export interface AssistantReply {
  content: string;
}

export interface CapabilityCallCandidate {
  activity_id: string;
  surface_version: string;
  capability_id: string;
  input_ref: string;
  effective_capability_ids?: string[];
  provider_replay?: ProviderToolCallReplay | null;
}

export interface ProviderToolCallReplay {
  provider_id: string;
  provider_model_id: string;
  provider_turn_id: string;
  provider_call_id: string;
  provider_tool_name: string;
  arguments: unknown;
  response_reasoning?: string | null;
  reasoning?: string | null;
  signature?: string | null;
}

/** `LoopModelUsage` — `cache_*` fields default 0 and are skipped when 0. */
export interface LoopModelUsage {
  input_tokens: number;
  output_tokens: number;
  cache_read_input_tokens?: number;
  cache_creation_input_tokens?: number;
}

// --- capability invocation ---

export interface ProviderToolCall {
  provider_id: string;
  provider_model_id: string;
  turn_id?: string | null;
  id: string;
  name: string;
  arguments: unknown;
  response_reasoning?: string | null;
  reasoning?: string | null;
  signature?: string | null;
}

export interface RegisterProviderToolCallRequest {
  tool_call: ProviderToolCall;
  activity_id?: string | null;
}

/** `CapabilityCallCandidate` is the response (same shape as above). */

export interface LoopRequest {
  activity_id: string;
  surface_version: string;
  capability_id: string;
  input_ref: string;
  approval_resume?: unknown | null;
  auth_resume?: unknown | null;
}

/** `Resolution` from `ironclaw_host_api::resolution` — externally tagged. */
export type Resolution =
  | { done: Outcome }
  | { denied: { deny: string; reason_kind?: string; summary?: string } }
  | { blocked: Partial<Record<"approval" | "auth" | "resource", GateWaypoint>> }
  | {
      suspended:
        | { process: { process: string; origin?: string } }
        | { dependent_run: { waypoint: GateWaypoint; result: unknown } }
        | { external_tool: GateWaypoint };
    };

export interface GateWaypoint {
  gate: string;
  origin?: string;
  resume?: string;
}

export interface Outcome {
  refs: OutcomeRefs;
  verdict: unknown;
  summary: string;
  progress?: unknown;
  terminate_hint?: unknown;
}

export interface OutcomeRefs {
  result: string;
  byte_len: number;
  preview?: string | null;
  preview_meta?: unknown;
  origin?: string | null;
  output_digest?: unknown;
}

// --- assistant drafts (host/transcript.rs) ---

export interface BeginAssistantDraft {
  reply: AssistantReply;
}

export interface UpdateAssistantDraft {
  message_ref: string;
  reply: AssistantReply;
}

export interface FinalizeAssistantMessage {
  reply: AssistantReply;
}

export interface AppendCapabilityResultRef {
  result_ref: string;
  safe_summary: string;
  provider_call?: ProviderToolCallReference | null;
  model_observation?: unknown | null;
  intrinsic_outcome?: unknown | null;
}

export interface ProviderToolCallReference extends ProviderToolCallReplay {
  capability_id: string;
}

// --- checkpoints (host/checkpoint.rs) ---

export type LoopCheckpointKind =
  "before_model" | "before_side_effect" | "before_block" | "final";

export interface StageCheckpointPayloadRequest {
  kind: LoopCheckpointKind;
  schema_id: string;
  /** Canonical payload bytes — JSON array of byte numbers (serde `Vec<u8>`). */
  payload: number[];
}

// ---------------------------------------------------------------------------
// Outcomes (loop_exit.rs) — serde `rename_all = "snake_case"`
// ---------------------------------------------------------------------------

export type LoopExit =
  | { completed: LoopCompleted }
  | { blocked: LoopBlocked }
  | { cancelled: LoopCancelled }
  | { failed: LoopFailed };

export interface LoopCompleted {
  completion_kind: LoopCompletionKind;
  reply_message_refs: string[];
  result_refs: string[];
  final_checkpoint_id: string | null;
  model_usage?: LoopModelUsage | null;
  exit_id: string;
}

export type LoopCompletionKind =
  | "final_reply"
  | "ask_user_reply"
  | "no_reply"
  | "delegated_result"
  | "result_only"
  | "nothing_to_report";

export interface LoopBlocked {
  kind: LoopBlockedKind;
  gate_ref: string;
  blocked_activity_id?: string | null;
  credential_requirements?: unknown[];
  checkpoint_id: string;
  state_ref: string;
  exit_id: string;
}

export type LoopBlockedKind =
  "approval" | "auth" | "resource" | "await_dependent_run" | "external_tool";

export interface LoopCancelled {
  reason_kind: "host_cancellation" | "host_interrupt";
  checkpoint_id: string | null;
  interrupted_message_refs: string[];
  exit_id: string;
}

export interface LoopFailed {
  reason_kind: LoopFailureKind;
  checkpoint_id: string | null;
  model_usage?: LoopModelUsage | null;
  exit_id: string;
  explanation_message_refs?: string[];
  safe_summary?: unknown | null;
}

export type LoopFailureKind =
  | "model_error"
  | "context_build_failed"
  | "capability_protocol_error"
  | "iteration_limit"
  | "invalid_model_output"
  | "checkpoint_rejected"
  | "checkpoint_unavailable"
  | "transcript_write_failed"
  | "driver_bug"
  | "interrupted_unexpectedly"
  | "no_progress_detected"
  | "policy_denied"
  | "compaction_unavailable"
  | "gate_not_supported"
  | "wall_clock_limit"
  | "model_call_limit"
  | "capability_invocation_limit";

export interface LoopWorkerFailure {
  kind: string;
  detail?: string | null;
}

/** `LoopWorkerOutcome` — externally tagged. */
export type LoopWorkerOutcome =
  { Exit: LoopExit } | { Failed: LoopWorkerFailure };

// ---------------------------------------------------------------------------
// Frames — externally tagged, mirroring `HostFrame` / `WorkerFrame`
// ---------------------------------------------------------------------------

/** `LoopCancellationSignal` (`host/progress.rs`). */
export interface LoopCancellationSignal {
  reason_kind: string;
  detail?: unknown;
}

export type HostFrame =
  | { Bootstrap: LoopWorkerBootstrap }
  | {
      HostResponse: {
        id: number;
        result: { Ok: unknown } | { Err: WireError };
      };
    }
  | { Cancel: LoopCancellationSignal }
  | "OutcomeAck";

export type WorkerFrame =
  | { HostRequest: { id: number; call: HostCall } }
  | { Outcome: LoopWorkerOutcome };

/** `WireError` — externally tagged. */
export type WireError =
  { Host: AgentLoopHostError } | { Compaction: unknown } | { Protocol: string };

/** `AgentLoopHostError` (`host/error.rs`). */
export interface AgentLoopHostError {
  kind: string;
  safe_summary: string;
  reason_kind?: unknown | null;
  gate_ref?: string | null;
  retry_after_ms?: number | null;
  next_fallback_index?: number | null;
  usage?: LoopModelUsage | null;
  detail?: string | null;
}

/** Discriminator for an `Err` wire response, for error handling. */
export function wireErrorSummary(error: WireError): string {
  if ("Host" in error) {
    return `host ${error.Host.kind}: ${error.Host.safe_summary}`;
  }
  if ("Compaction" in error) {
    return `compaction: ${JSON.stringify(error.Compaction)}`;
  }
  return `protocol: ${error.Protocol}`;
}
